Zend/zend_execute.c

```diff
@@ -1338,7 +1514,29 @@ static zend_never_inline void zend_assign_op_overloaded_property(zval *object, z
 		zptr = z;
 		ZVAL_DEREF(z);
 		SEPARATE_ZVAL_NOREF(z);
-		binary_op(z, z, value);
+		
+		if (UNEXPECTED(ZEND_OBJECT_HAS_TYPE_HINTS(object))) {
+			zend_property_info *prop_info = zend_object_fetch_property_type_info(object, property, cache_slot);
+
+			if (prop_info) {
+				zval tmp;
+
+				ZVAL_UNDEF(&tmp);
+				binary_op(&tmp, z, value);
+				if (!zend_verify_property_type(prop_info, &tmp, ZEND_CALL_USES_STRICT_TYPES(EG(current_execute_data)))) {
+					zend_verify_property_type_error(prop_info, Z_STR_P(property), &tmp);
+					OBJ_RELEASE(Z_OBJ(obj));
+					return;
+				}
+
+				ZVAL_COPY_VALUE(z, &tmp);
```

Comment:
This code looks wrong. You don't free original value of property and lose original value or reference counter.
The following script crashes:

class A {
  public string $x;
}
$a = new A;
$a->x = "a";
$a->x .= "b";
$a->x .= "c";